### PR TITLE
Make sure when on last block focus cannot enter the block.

### PR DESCRIPTION
**Pull request by alexstine**
## Description

Whenever you are in Gutenberg Navigation Mode, I found a bug that effects how next focus works. More details in testing instructions below.

## How has this been tested?

Tested in Firefox with NVDA screen reader.

### Testing Instructions
1. Open a Post or Page.
2. Insert some Paragraph Blocks followed by a Block with a button such as a Columns block.
3. Enter Gutenberg Navigation Mode.
4. Tab through your Paragraph Blocks.
5. Once you arrive at the Columns Block, Tab once more. Instead of focus being placed on the first button of the Block Variations panel, focus will be placed on the Post tab of the Settings Sidebar.

## Types of changes

Bug fix

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). 
- [x] I've updated related schemas if appropriate.

### CODE REVIEW

File: packages/edit-post/src/components/sidebar/settings-header/index.js [hunk 1]
```diff
@@ -67,6 +68,7 @@ const SettingsHeader = ( { sidebarName } ) => {
 						className={ `edit-post-sidebar__panel-tab ${ templateActiveClass }` }
 						aria-label={ templateAriaLabel }
 						data-label={ __( 'Template' ) }
+						data-focus-at-end
```
Comment (alexstine) [inline packages/edit-post/src/components/sidebar/settings-header/index.js]: Not sure if this is needed. Not familiar with how Template Mode works.

File: packages/block-editor/src/components/block-tools/block-selection-button.js [hunk 2]
```diff
@@ -196,6 +196,15 @@ function BlockSelectionButton( { clientId, rootClientId, blockElement } ) {
 				if ( navigateDown ) {
 					nextTabbable = focus.tabbable.findNext( blockElement );
 
+					if ( null === getNextBlockClientId( clientId ) ) {
```
Comment (alexstine) [inline packages/block-editor/src/components/block-tools/block-selection-button.js]: Couldn't think of a better way to do this. Open to feedback on how to better focus the Settings Sidebar. This code was great before these new Blocks, it just doesn't work with the more dynamic Blocks.

**Review by alexstine — COMMENTED** [review]
Are there other packages to apply this to? The Site Editor and other sections are still a bit too inaccessible for me to use at this point so not sure if this could be a problem elsewhere.

**Comment by youknowriad** [comment]
So just to clarify, the problem here is that we get stuck tabbing between the last block and the "button" inside it right instead of going to the sidebar of the editor?

It feels to me the solution here is not to add a specific element to tab into like the PR tries to do with `data-focus-at-end` (because that element can change over time if say we change the design of the sidebar or add things in the footer) but instead we should ignore the elements in the block list somehow.

Basically I think the fix should be somewhere in these lines https://github.com/WordPress/gutenberg/pull/37965/files#diff-11abd7881539ce65efd801429f377af4876374040aef15b4255bc96658b790c2R208-R212 where we should find the next tubbable after the last block (but not after the last block's label which include elements inside the block itself)

**Comment by alexstine** [comment]
@youknowriad Yes, that is correct. You get stuck there and you can only move backwards.

How exactly could I determine the next element would be inside a block? It seems difficult to determine as there are not really any unique classes I could ignore. Maybe trying to detect the parent element? Either way, there's going to be some gymnastics involved. I really don't favor my solution either but I am not sure we should be reliant on classes to tell us if focus would be in block content or not. 😕

**Comment by alexstine** [comment]
@youknowriad I fixed it but the solution is super ugly. Let me know your thoughts about this approach.
1. I kept the check to make sure I only fire this code on last block. Then I checked to make sure focus element was found.
2. Then I got the current block with constructing the id with clientId.
3. Then I started a loop. If the focus element was inside the block ID as a child element, keep finding the next focus element until I am no longer in the block.

I think it was a rather interesting idea of mine but not sure how reasonable it is as there's a lot of DOM interaction and checking.

File: packages/block-editor/src/components/block-tools/block-selection-button.js [hunk 3]
```diff
@@ -196,6 +196,28 @@ function BlockSelectionButton( { clientId, rootClientId, blockElement } ) {
 				if ( navigateDown ) {
 					nextTabbable = focus.tabbable.findNext( blockElement );
```
Comment (youknowriad) [inline packages/block-editor/src/components/block-tools/block-selection-button.js]: I've been looking at this PR again and for me the root cause is in this line basically what we're saying here is:

If there's no next block, focus the next tabbable element after the current block element, the problem is that this doesn't exclude the content of the current block (because it comes after blockElement), so I think a proper if is just to continue calling "findNext" until we're out of the current block.

Something like this maybe?

```
nextTabbable = blockElement;
					do {
						nextTabbable = focus.tabbable.findNext( nextTabbable );
					} while ( nextTabbable && blockElement.contains( nextTabbable ) );

```

Comment (alexstine) [reply packages/block-editor/src/components/block-tools/block-selection-button.js]: I am glad you knew what I was trying to do. 👍 This works perfectly. 🙇‍♂️ Thanks for suggestion.

**Review by youknowriad — APPROVED** [review]

**Comment by alexstine** [comment]
Waiting for Accessibility team to check it.

https://wordpress.slack.com/archives/C02RP4X03/p1642512823187900

I am hoping I can find at least one other reviewer just to make sure I haven't missed anything here as far as accessibility goes.

**Review by tellthemachines — APPROVED** [review]
Tested this with VoiceOver on Safari and Firefox; it's working well ✅

**Review by kevin940726 — APPROVED** [review]
Code makes sense to me, thanks 👍 !
